File: go/linkedlist/doubly_linkedlist/doubly_linkedlist.go
package doubly_linkedlist

import "fmt"

//双向链表的实现

//节点定义
type ListNode struct {
	val        interface{}
	prev, next *ListNode
}

//新建节点
func newNode(val interface{}) *ListNode {
	return &ListNode{val, nil, nil}
}

//链表定义
type LinkedList struct {
	head *ListNode
	size int
}

//新建链表
func New() *LinkedList {
	return &LinkedList{nil, 0}
}

//插入数据到链表头部
func (lis *LinkedList) PushFront(val interface{}) *ListNode {
	node := newNode(val)
	if lis.head != nil {
		lis.head.prev = node
		node.next = lis.head
	}

	lis.head = node
	lis.size++
	return node
}

//插入数据到链表尾部
func (lis *LinkedList) PushBack(val interface{}) *ListNode {
	if lis.head == nil {
		return lis.PushFront(val)
	}

	node := newNode(val)
	p := lis.head
	for p.next != nil {
		p = p.next
	}

	p.next = node
	node.prev = p
	lis.size++
	return node
}

//在某个节点之后插入数据
func (lis *LinkedList) PushAfter(p *ListNode, val interface{}) *ListNode {
	if p == nil {
		return nil
	}

	next := p.next
	node := newNode(val)
	node.next = next
	p.next = node
	node.prev = p
	if next != nil {
		next.prev = node
	}
	lis.size++
	return node
}

//在某个节点之前插入元素
func (lis *LinkedList) PushBefore(p *ListNode, val interface{}) *ListNode {
	if p == nil {
		return nil
	}

	node := newNode(val)
	prev := p.prev
	if prev == nil {
		lis.PushFront(val)
	} else {
		p.prev = node
		node.next = p
		prev.next = node
		node.prev = prev
		lis.size++
	}
	return node
}

//删除节点
func (lis *LinkedList) Delete(p *ListNode) {
	if p == nil || lis.head == nil {
		return
	}

	//删除的是头节点
	if p == lis.head {
		lis.head = p.next
	} else {
		prev, next := p.prev, p.next
		prev.next = next
		if next != nil {
			next.prev = prev
		}
	}
	lis.size--
}

//根据值查找节点
func (lis *LinkedList) Find(val interface{}) *ListNode {
	if lis.head == nil {
		return nil
	}

	p := lis.head
	for p != nil && p.val != val {
		p = p.next
	}
	return p
}

func (lis *LinkedList) Size() int {
	return lis.size
}

//打印所有链表数据
func (lis *LinkedList) PrintData() {
	for p := lis.head; p != nil; p = p.next {
		fmt.Print(p.val, " ")
	}
	fmt.Println()
}

File: python/linkedlist/doubly_linked_list.py

# 双向链表实现

class ListNode:

    def __init__(self, val):
        self.val = val
        self.prev = None
        self.next = None


class LinkdeList:

    def __init__(self):
        self.head = None

    # 查找元素
    def find(self, val) -> ListNode:
        p = self.head
        while p and p.val != val:
            p = p.next
        return p

    # 添加节点至链表尾部
    def add(self, val) -> ListNode:
        node = ListNode(val)
        if not self.head:
            self.head = node
        else:
            p = self.head
            while p.next:
                p = p.next
            p.next = node
            node.prev = p

        return node

    # 在某个节点之后添加节点
    def addAfter(self, p: ListNode, val: int):
        if not p:
            return None
        node = ListNode(val)
        n = p.next
        node.next = p.next
        p.next = node
        node.prev = p
        if n:
            n.prev = node
        return node

    # 在某个节点之前插入节点
    def addBefore(self, p: ListNode, val: int):
        if not p:
            return None

        node = ListNode(val)
        if p == self.head:
            node.next, p.prev = p, node
            self.head = node
        else:
            prev = p.prev
            node.next, p.prev = p, node
            prev.next, node.prev = node, prev
        return node

    # 删除链表中的节点
    def delete(self, p: ListNode):
        if not p:
            return
        if p == self.head:
            self.head = p.next
        else:
            prev, back = p.prev, p.next
            prev.next = p.next
            if back:
                back.prev = prev

    # 打印所有的数据
    def print(self):
        p = self.head
        while p:
            print(p.val, end=' ')
            p = p.next
        print()


if __name__ == '__main__':
    lis = LinkdeList()
    n1 = lis.add(1)
    lis.add(5)
    n3 = lis.add(3)

    n10 = lis.addAfter(n3, 10)
    n20 = lis.addAfter(n10, 20)

    lis.print()

    lis.addBefore(n20, 15)
    h = lis.addBefore(n1, -1)
    lis.print()

    lis.delete(n1)
    lis.delete(h)
    lis.delete(n20)
    lis.print()
